### PR TITLE
[Experimental] Product filters/overlay nav block redesign

### LINKED ISSUE #50273: Product Filters Redesign: Update Overlay Navigation UX
@dinhtungdu and @jarekmorawski discussed the UX around Overlay Navigation block, which is currently following this specification: https://github.com/woocommerce/woocommerce/pull/48732#issuecomment-2225223051. From the conversation, we agreed to change the behavior to follow what Core does with the No Results block. Here is the new specification for the Overlay Navigation block settings, in the context of the Product Filters:

- The Overlay Navigation block is hidden (completely removed) when the Overlay setting is set to Never.
- The Overlay Navigation block is visible when the Overlay setting is set to Mobile or Always.
- The block settings are always belong to the Overlay Navigation block.
- The block description of the Overlay Navigation should be update to inform user about its behavior with different Overlay setting.

### PR DESCRIPTION
**Pull request by roykho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/50273

- ~I've updated the Overlay Navigation block icon to a generic button icon because we're also using this block as a variation for the trigger button and it didn't seem correct to show an icon of "close" when it actually "opens".~ Each overlay navigation open/close will have it's own unique icon.
- I've implemented the logic to remove the overlay navigation block when the overlay setting is set to "Never".
- I've implemented the logic to retain the user settings when they change their mind and set the overlay setting to "Mobile" or "Always".
- I've implemented the logic to add the overlay navigation block back when the overlay setting is set to "mobile" and "always".
- ~I've removed the overlay button styling from the parent Product Filters (Experimental) block since these settings are already there on the Overlay Navigation (Experimental) block.~ I've moved the icon picker to the "Open overlay navigation only".
- Note that we should probably remove the Navigation block (open overlay) from the Product Filters Overlay as it serve no purpose inside there.
- One thing I didn't do in this PR is update the overlay navigation block description depending on the overlay setting. This was mentioned in the original issue however I am not 100% sure what this means. Please clarify if you know. I can either update this PR or I can add it in a separate PR.
- ~I will add tests once we agree this is the direction to take.~ Tests have been added.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

> [!NOTE]
> **Test by developers only**
>
> **Pre-req:**
> Ensure you have the WC Beta Tester Plugin installed and go to Tools > WCA Test Helper > Features. Toggle on the experimental-blocks option.

1. Go to the site editor and navigate to Patterns->General template parts->Product Filters (Experimental). Make sure it has been reseted.
2. Ensure you don't see the Overlay Navigation block in the list view and on the canvas.
3. Select the parent Product Filters (Experimental) and in the inspector controls > Overlay change it to Mobile.
4. Ensure you now see the Overlay Navigation block added in the list view and on the canvas.
5. Test step 3 again but with Always. You should also see the same as step 4.
6. Select the Overlay Navigation (Experimental) block and make some changes to the inspector controls Style.
7. Select the parent Product Filters (Experimental) and in the inspector controls > Overlay change it to Never.
8. Ensure the Overlay Navigation block is now removed.
9. Select the parent Product Filters (Experimental) and in the inspector controls > Overlay change it to Mobile.
10. Ensure you now see the Overlay Navigation block added in the list view and on the canvas and that all the styles you made before in step 6 is retained.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 
Product Filters: update overlay navigation UX
</details>

### CODE REVIEW

**Comment by roykho** [comment]
~Note that I will add e2e tests once we have reviewed this PR and this direction is the way we want.~ Tests have been added.

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
Thanks for working on this PR! I tested against a fresh JN site and I can confirm the button is working as expected following your testing instructions.

> I've removed the overlay button styling from the parent Product Filters (Experimental) block since these settings are already there on the Overlay Navigation (Experimental) block.

I haven't done the code review yet because we're missing the icon chooser for the open button (not the close one in the Filter Overlay).

image

> I've updated the Overlay Navigation block icon to a generic button icon because we're also using this block as a variation for the trigger button and it didn't seem correct to show an icon of "close" when it actually "opens".

Can you clarify this? Are you saying that you're using the same generic button icon for both close and open buttons? Variations can have different icons per variation, so I don't get what you mean.

> One thing I didn't do in this PR is update the overlay navigation block description depending on the overlay setting. This was mentioned in the original issue however I am not 100% sure what this means. Please clarify if you know. I can either update this PR or I can add it in a separate PR.

When the overlay setting is Mobile, the overlay navigation block is always visible. I thought we could use the block description to inform users about that. But it may not be possible. You can leave that one out, we can add some help text later if needed.

In addition, I can't see the close button inside the Filter Overlay template part anymore:
image

**Comment by roykho** [comment]
@dinhtungdu Thanks for the review:

> I haven't done the code review yet because we're missing the icon chooser for the open button (not the close one in the Filter Overlay).

I have added the icon picker to the Open overlay navigation https://github.com/woocommerce/woocommerce/pull/51211/commits/770c28db0e27226885c633d76cd31ea8adb83350

> Can you clarify this? Are you saying that you're using the same generic button icon for both close and open buttons? Variations can have different icons per variation, so I don't get what you mean.

So initially I tried using different icons and it didn't work when I declared it in the block variations. Then I found the missing piece was that we had to declared `isActive`. After I did that, it started working. https://github.com/woocommerce/woocommerce/pull/51211/commits/1b888bd16eea0841cb785e7e52fb667ef3d28873

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/overlay-navigation/inspector-controls.tsx
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/overlay-navigation/inspector-controls.tsx:109]: The label is missing here.

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
Thanks for the update! Mostly looks good to me. Apart from the review comment, can you also fix the spacing between style and button settings (attached below). The tests are also failing, can you look at them?

image

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/overlay-navigation/inspector-controls.tsx
Comment (roykho) [reply plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/overlay-navigation/inspector-controls.tsx:109]: Thanks fixed https://github.com/woocommerce/woocommerce/pull/51211/commits/0dfae59a1e4d8022454f8ef41169bbcfc6cb04f2

**Comment by roykho** [comment]
@dinhtungdu thanks for the review.

> can you also fix the spacing between style and button settings (attached below).

That was also because of the missing label which I've added https://github.com/woocommerce/woocommerce/pull/51211/commits/0dfae59a1e4d8022454f8ef41169bbcfc6cb04f2

> The tests are also failing, can you look at them?

I have fixed the test and also skip a test within the overlay context because once we overhaul the re-design there, we will have to change the tests too. I've also added new tests to cover this PR.

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx
Comment (thealexandrelara) [inline plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx]: I believe it would be better to select only the blocks that are inner blocks of the current Product Filters block instance. And also, instead of selecting them by the position index, perhaps it would be better to filter them by the `attributes.triggerType === 'open-overlay` instead. This way, we can prevent the error that I mentioned in my testing results.

**Review by thealexandrelara — CHANGES_REQUESTED** [review]
Thank you for working on this, Roy. I left a comment in the code that is connected to the testing scenario eight listed below. 

## Test summary:

✅ Scenario 1: Remove the Overlay Navigation block when the overlay setting is set to "Never"
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The Product Filters (Experimental) has never been modified or it has been resetted to its initial state.
- Expected Outcome: The Overlay Navigation block should not appear on the canvas or in the list view.
- Details: When the template part is loaded, no Overlay Navigation block is displayed. By opening the List View section, the block is also not visible there.

---

✅ Scenario 2: Add the Overlay Navigation block when the overlay setting is changed from "Never" to "Mobile"
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Never".
- Expected Outcome: The Overlay Navigation block should appear on the canvas and in the list view.
- Details: Once the merchant sets the Overlay setting to "Mobile" the Overlay Navigation block appears on the canvas and in the list view.

---

✅ Scenario 3: Add the Overlay Navigation block when the overlay setting is changed from "Never" to "Always"
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Never".
- Expected Outcome: The Overlay Navigation block should appear on the canvas and in the list view.
- Details: Once the merchant sets the Overlay setting to "Always" the Overlay Navigation block appears on the canvas and in the list view.

---

✅ Scenario 4: Persist block customizations when changing from "Mobile" to "Never" and then back to "Mobile"
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Always".
    - The merchant customize the block by changing its attributes like font size, color, border and style.
    - The merchant set the "Overlay" setting to "Never".
    - The merchant set the "Overlay" setting back to "Mobile".
- Expected Outcome: The Overlay Navigation block should re-appear with the previously customized settings.
- Details: Once the merchant set the "Overlay" setting back to "Mobile.", the block appears with all the customizations made before changing that setting to "Never". This behaviour can be noticed even after a full page reload.

---

✅ Scenario 5: Persist block customizations when changing from "Always" to "Never" and then back to "Always"
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Always".
    - The merchant customize the block by changing its attributes like font size, color, border and style.
    - The merchant set the "Overlay" setting to "Never".
    - The merchant set the "Overlay" setting back to "Always".
- Expected Outcome: The Overlay Navigation block should re-appear with the previously customized settings.
- Details: Once the merchant set the "Overlay" setting back to "Always", the block appears with all the customizations made before changing that setting to "Never". This behaviour can be noticed even after a full page reload.

---

✅ Scenario 6: Persist block customizations when changing from "Always" to "Never" 
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Always".
    - The merchant set the "Overlay" setting to "Never".
- Expected Outcome: The Overlay Navigation block should be removed from the canvas and the list view.
- Details: Once the merchant set the "Overlay" setting to "Never", the block is removed from the canvas and the list view.

---

✅ Scenario 7: Persist block customizations when changing from "Mobile" to "Never" 
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Mobile".
    - The merchant set the "Overlay" setting to "Never".
- Expected Outcome: The Overlay Navigation block should be removed from the canvas and the list view.
- Details: Once the merchant set the "Overlay" setting to "Never", the block is removed from the canvas and the list view.

---

❌  Scenario 8: Should keep the Close button on the Product Filters Overlay template part.
- Conditions:
    - A merchant visits the Filters Overlay template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Always" or "Mobile".
    - The merchant set the "Overlay" setting to "Never".
- Expected Outcome: The Overlay Navigation block which contains the Close button should NOT be removed from the canvas or te list view.
- Details: The "Close" button is being removed after the user changes the "Overlay" setting from "Always" or "Mobile" to "Never"

https://github.com/user-attachments/assets/22825c9a-a334-49b8-9bca-82577a6f8dc6

---

**Comment by roykho** [comment]
@thealexandrelara thanks for the review.

> ❌ Scenario 8: Should keep the Close button on the Product Filters Overlay template part.

This is expected. The overlay template part needs a rework as I mentioned in the PR. Though with revised logic https://github.com/woocommerce/woocommerce/pull/51211/commits/5bfd176e354932f98d5ca5627adc0469213fb85d this consequently fix that issue though we still need to rework the overlay template part. For example the overlay navigation (open overlay) block should not show in the list nor the editor canvas.

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx
Comment (roykho) [reply plugins/woocommerce-blocks/assets/js/blocks/product-filters/edit.tsx]: Good idea. I've revise the logic https://github.com/woocommerce/woocommerce/pull/51211/commits/5bfd176e354932f98d5ca5627adc0469213fb85d

**Comment by roykho** [comment]
@thealexandrelara @dinhtungdu - thanks for the reviews so far. I have pushed up the changes requested. Another round please.

**Review by thealexandrelara — APPROVED** [review]
After re-reviewing the PR I verified that all of the testing scenarios are currently working as expected. Thank you for fixing the raised issue.

✅ Scenario 8: Should keep the Close button on the Product Filters Overlay template part.

- Conditions:
    - A merchant visits the Filters Overlay template part.
    - The "Overlay" setting from the Product Filters (Experimental) block is initially set to "Always" or "Mobile".
    - The merchant set the "Overlay" setting to "Never".

- Expected Outcome: The Overlay Navigation block which contains the Close button should NOT be removed from the canvas or te list view.

- Details: The "Close" button is not being removed anymore, only the Overlay Navigation block within the Product Filters block is removed.

**Comment by dinhtungdu** [comment]
> > can you also fix the spacing between style and button settings (attached below).

> That was also because of the missing label which I've added https://github.com/woocommerce/woocommerce/commit/0dfae59a1e4d8022454f8ef41169bbcfc6cb04f2

@roykho I still see the issue with your latest commit 
image

**Review by dinhtungdu — APPROVED** [review]
Thanks for the updates. The PR looks good to me now. I'm pre-approving this so it won't block merging. But please fix [this](https://github.com/woocommerce/woocommerce/pull/51211#issuecomment-2342731177) before merging this PR.

**Comment by roykho** [comment]
@dinhtungdu  I am not seeing the issue you're seeing. This is what I see:
Google Chrome - Filters Overlay ‹ Template part ‹ woo-blocks-dev ‹ Editor — WordPress 2024-09-11 at 5 40 51 AM

**Comment by roykho** [comment]
@thealexandrelara - Do you also see the [spacing issue](https://github.com/woocommerce/woocommerce/pull/51211#issuecomment-2342731177) observed by Tung?

**Comment by thealexandrelara** [comment]
> @thealexandrelara - Do you also see the [spacing issue](https://github.com/woocommerce/woocommerce/pull/51211#issuecomment-2342731177) observed by Tung?

I'm not seeing the spacing issue. I create a new JN site to test and here is what I see:

CleanShot 2024-09-11 at 15 59 30@2x

**Pull request merged**
